### PR TITLE
129 add MLP

### PR DESCRIPTION
**Pull request by nmaarnio**
Added an MLP implementation that uses Keras/Tensorflow.

Includes:
- Separated classifier and regressor implementations
- Checks for inputs
- Docs
- Tests

When PR #210 is merged, the private functions of this PR (input check and keras optimizer getter) can be moved to `model_utils`. 

Currently, probabilistic MLP is not implemented. This can be done in a later PR, on top of this implementation.

Cross-validation is not implemented either.

There might be some mistakes in the allowed parameters/parameter combinations, so it would be good if @msmiyels or someone else could try to spot these. Also, the exposed parameter space can be expanded or shrunk as you see fit.

### CODE REVIEW

File: eis_toolkit/prediction/mlp.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+    last_activation: Literal["softmax", "sigmoid"] = "softmax",
+    epochs: int = 50,
+    batch_size: int = 32,
+    optimizer: Literal["adam"] = "adam",
+    learning_rate: Number = 0.001,
+    loss_function: Literal["categorical_crossentropy", "binary_crossentropy"] = "categorical_crossentropy",
+    dropout_rate: Optional[Number] = None,
+    early_stopping: bool = True,
+    es_patience: int = 5,
+    metrics: Optional[Sequence[Literal["accuracy", "precision", "recall"]]] = ["accuracy"],
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:81]: F1score does not seem to be listed at least here https://keras.io/api/metrics/

File: eis_toolkit/prediction/mlp.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+
+@beartype
+def train_MLP_classifier(
+    X: np.ndarray,
+    y: np.ndarray,
+    neurons: Sequence[int] = [16],
+    validation_split: Optional[float] = 0.2,
+    activation: Literal["relu"] = "relu",
+    output_neurons: int = 1,
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:71]: For binary classification, is not one class what we want? That's what I've understood. Related to this, should we have `sigmoid` as the default last activation and `binary_crossentropy` as the loss function if we make binary classification the default mode?

File: eis_toolkit/prediction/mlp.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+    )
+
+    return model, history.history
+
+
+@beartype
+def train_MLP_regressor(
+    X: np.ndarray,
+    y: np.ndarray,
+    neurons: Sequence[int] = [16],
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:166]: Do you suggest that we remove the default for this parameter completely? I guess at least for QGIS plugin users we could offer a choice to with this `n * 2 + 1` default which would be calculated from the inputs.

File: tests/prediction/mlp_test.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    assert "loss" in history, "History should contain loss"
+    assert "mse" in history, "History should contain mse"
+    assert len(model.layers) > 0, "Model should have layers"
+
+    scores = model.evaluate(X_test, y_test, verbose=0)
+    test_mse = scores[1]
+
+    # Can't check if output is equal to a single number as there is some randomness involved in creating MLP model
+    assert history["mse"][-1] < 0.3, "Last epoch's MSE should be reasonably low"
+    assert test_mse < 0.3, "Test MSE should be reasonably low"
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply tests/prediction/mlp_test.py:53]: True, good catch!

File: eis_toolkit/prediction/mlp.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+
+    # 3. Train the model
+    # Early stopping callback
+    callbacks = [keras.callbacks.EarlyStopping(monitor="val_loss", patience=es_patience)] if early_stopping else []
+
+    history = model.fit(
+        X,
+        y,
+        epochs=epochs,
+        validation_split=validation_split if validation_split else 0,
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:252]: Yeah good idea. I added this too. Do you think we want to include this option to provide already split validation data for the Sklearn ML models too?

File: eis_toolkit/prediction/mlp.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    # 2. Create and compile a sequential model
+    model = keras.Sequential()
+
+    for neuron in neurons:
+        model.add(keras.layers.Dense(units=neuron, activation=activation))
+
+        if dropout_rate is not None:
+            model.add(keras.layers.Dropout(dropout_rate))
+
+    model.add(keras.layers.Flatten())
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py]: Okay, thanks. I think I just copied the Flatten layer from Luca's code where I saw it – perhaps it got accidentally there. I removed it now and we could assume that people who use EIS for scripting / notebooks can prepare their data. We could add some sort of check for the input shape if you think that's a good idea. For plugin users, we can make a reshaping/preparation function that we call in the CLI before passing data to these functions.

**Comment by nmaarnio** [comment]
Hi Micha!

Thanks for the review. I made corrections/additions already for most of the issues, and for a few I had a question/comment. We can have a chat if there's still further to discuss about these points, and even if not we could have another modeling meet in January

File: eis_toolkit/prediction/mlp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:88]: Yeah, I was puzzled why I couldn't find in the docs since it should be a used metric. Odd that it didn't show in the site I looked, but I added it now!

File: eis_toolkit/prediction/mlp.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:78]: Okay, got it :+1: . Sounds like we have two valid options, I will go for the 1 output class, binary crossentropy, sigmoid combo now if you think that's okay.

File: eis_toolkit/prediction/mlp.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:273]: Yes, I was just thinking that we could provide an option in the functions we offer to give a separate validation dataset. When defined, it would override calling the `split_data` inside  `_train_and_validate_sklearn_model` and directly assign `X_valid` and `y_valid` from the given dataset. But I don't see a need to add this, just thought I'd ask. Anyway it is separate from this MLP issue.

**Comment by nmaarnio** [comment]
Hi,

I finished implementing the changes you requested, so should be ready for another round of review!

File: eis_toolkit/prediction/mlp.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    if batch_size <= 0:
+        raise exceptions.InvalidParameterValueException("Batch size must be greater than 0.")
+
+    if epochs <= 0:
+        raise exceptions.InvalidParameterValueException("Number of epochs must be greater than 0.")
+
+    if output_neurons <= 0:
+        raise exceptions.InvalidParameterValueException("Number of output neurons must be greater than 0.")
+
+    if output_neurons > 1 and loss_function == "binary_crossentropy":
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:65]: Yes, good idea. But is `output_neurons` == 2 then invalid in all cases, or am I missing something?

File: eis_toolkit/prediction/mlp.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+    for neuron in neurons:
+        model.add(keras.layers.Dense(units=neuron, activation=activation))
+
+        if dropout_rate is not None:
+            model.add(keras.layers.Dropout(dropout_rate))
+
+    model.add(keras.layers.Dense(units=output_neurons, activation=last_activation))
+
+    model.compile(
+        optimizer=_keras_optimizer(optimizer, learning_rate=learning_rate), loss=loss_function, metrics=metrics
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:295]: This is good to know. I would imagine there are some drawbacks in using legacy optimizers in general? I am thinking we could detect the OS where the tool is run and based on that select legacy/non-legacy optimizer, but do you think this feature should be added already? I would personally lean towards making a note of this and doing a few tests in the upcoming months to see the effects, before making this change

File: eis_toolkit/prediction/mlp.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+    neurons: Sequence[int],
+    validation_split: Optional[float] = 0.2,
+    validation_data: Optional[Tuple[np.ndarray, np.ndarray]] = None,
+    activation: Literal["relu", "linear", "sigmoid", "tanh"] = "relu",
+    output_neurons: int = 1,
+    last_activation: Literal["linear"] = "linear",
+    epochs: int = 50,
+    batch_size: int = 32,
+    optimizer: Literal["adam", "adagrad", "rmsprop", "sdg"] = "adam",
+    learning_rate: Number = 0.001,
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:219]: Since it can be misleading/confusing if the default `learning_rate` is substituted silently in the background if `sdg` is chosen as the optimizer, I think we could leave it like this. I suppose it is reasonable to assume that if the user changes the optimizer, they would also tinker other parameters and have an understanding of what they're doing?

File: eis_toolkit/prediction/mlp.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:65]: Since you included `sparse_categorical_crossentropy` in this check, I assumed it should also be added as a parameter value for `loss_function` in `train_MLP_classifier` ;). Let me know if this wasn't the idea

File: tests/prediction/mlp_test.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+X_DIABETES, Y_DIABETES = load_diabetes(return_X_y=True)
+
+
+def test_train_MLP_classifier():
+    """Test that training MLP classifier works as expected."""
+    X = StandardScaler().fit_transform(X_IRIS)
+    y = one_hot_encode(Y_IRIS).toarray()
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
+
+    model, history = train_MLP_classifier(
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply tests/prediction/mlp_test.py:20]: Ah, forgot to modify the tests, good catch :+1:

File: eis_toolkit/prediction/mlp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:295]: Ok, got it. Good if there are no drawbacks and we can simply switch to the legacy versions. I don't really have an opinion then, whether to change now or when it is seen necessary, so up to you

File: eis_toolkit/prediction/mlp.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py:65]: Okay :+1: . Again I would let you decide whether to add `sparse_categorical_crossentropy` or not. We have one-hot enconding available for users and can even perform it automatically (inside MLP or inside CLI for plugin users), if we want. I would just add to the docstring that we expect one-hot encoded data if that is the case, or that it depends on the loss function if that is the case.

**Comment by nmaarnio** [comment]
Ready for another round :mag:

File: eis_toolkit/prediction/mlp.py [hunk 11]
```diff
@@ -3,7 +3,7 @@
 
 import numpy as np
 from beartype import beartype
-from keras.optimizers import SGD, Adagrad, Adam, RMSprop
+from keras.optimizers.legacy import SGD, Adagrad, Adam, RMSprop
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/mlp.py]: Great, thanks!
